src/main/java/io/cdap/plugin/SFTPCopyAction.java

```diff
@@ -151,53 +152,85 @@ public void run(ActionContext context) throws Exception {
       }
       copySFTPFiles(fileSystem, destination, sftpConnector, context);
     } catch(Exception e) {
-      throw new RuntimeException(String.format("Error occurred while connecting to SFTP Server %s", e.getMessage(), e));
+      throw new RuntimeException(String.format("Error occurred while copying files: %s", e.getMessage()), e);
     } finally {
         if (sftpConnector != null) {
           sftpConnector.close();
         }
     }
   }
 
-  private void copySFTPFiles(FileSystem fileSystem, Path destination, SFTPConnector SFTPConnector,
+  /**
+   * copySFTPFiles recursively copies all files and subdirectories in a given directory.
+   *
+   * @param fileSystem The destination file system to copy to
+   * @param destination The destination base path to copy to
+   * @param connector The SFTP connector to use
+   * @param context The context for the action
+   * @throws SftpException If any SFTP errors occur
+   */
+  private void copySFTPFiles(FileSystem fileSystem, Path destination, SFTPConnector connector,
                             ActionContext context) throws SftpException, IOException {
-    ChannelSftp channelSftp = SFTPConnector.getSftpChannel();
-    Vector files = channelSftp.ls(config.getSrcDirectory());
-    List<String> filesCopied = new ArrayList<>();
-    for (int index = 0; index < files.size(); index++) {
-      Object obj = files.elementAt(index);
+    ChannelSftp channel = connector.getSftpChannel();
+    ArrayList<String> filesCopied = new ArrayList<>();
+    Vector entries = channel.ls(config.getSrcDirectory());
+    for (Object obj : entries) {
       if (!(obj instanceof ChannelSftp.LsEntry)) {
         continue;
       }
       ChannelSftp.LsEntry entry = (ChannelSftp.LsEntry) obj;
-      if (".".equals(entry.getFilename()) || "..".equals(entry.getFilename())) {
-        // ignore "." and ".." files
-        continue;
+      recursiveSFTPCopy(channel, fileSystem, Paths.get(config.getSrcDirectory()), destination, filesCopied, entry);
```

Comment:
this is a change in behavior. Somebody who was relying on a non-recursive copy may now have unexpected behavior.

Instead, we should introduce a new property for whether files should be read recursively, similar to what we have in the File sources. Since it's a new property, it should default to false to be backwards compatible.

Reply:
The original behavior of this plugin was to fail if it encountered a folder. Since it never worked with folders in the first place, it probably isn't really backwards incompatible with existing behavior right?

Reply:
I see, it's not incompatible then